### PR TITLE
Port AffineSystem to SystemScalarConverter

### PR DESCRIPTION
**Pull request by jwnimmer-tri**
This is an illustration of how scalar-converting copy constructors operate across type hierarchies.  I believe these are currently the only two `System`s in the tree that use both inheritance and scalar conversion at the same time.

---
This change is [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/6743)

### CODE REVIEW

**Comment by jwnimmer-tri** [comment]
+@sherm1 for feature review, please.  This relates to `SetSystemScalarConverter` that we discussed in #6684, so you should probably own the review here even if it sits a little while in your TODO queue.

**Comment by sherm1** [comment]
Feature review pass done. I am trying to infer the general instructions for writing extensible, transmogrifiable System classes and not sure I've got it yet. I think it is something like: add a protected constructor whose first argument is a SystemScalarConverter, and (optionally?) a sugar constructor whose first argument is a `SystemTypeTag<S>` which will delegate to the first one. VectorSystem has both constructors but TimeVaryingAffineSystem has just the first one. AffineSystem has neither, suggesting it can't be extended(?). In that case LinearSystem (which derives from AffineSystem) can't be tranmogrified now. Is that right?

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-KqYgeGG1xWXLLDji5HF:bjses4e)*

**Comment by jwnimmer-tri** [comment]
> [instructions are like] ... add a protected constructor whose first argument is a SystemScalarConverter, and (optionally?) a sugar constructor whose first argument is a SystemTypeTag<S> which will delegate to the first one.

That's a reasonable proposal and a good first answer.

Additionally, you could choose to offer an overloaded ctor with no transmogrification arguments of any kind, and then delegate to the SystemScalarConverter overload with the converter set to a default-constructed object.  In this PR,`VectorSystem` does this (by keeping the pre-existing, transmogrification-unaware constructor intact).

Alternatively, if the protected constructor body is written (defined) in the header file and you want to _mandate_ that all subclasses transmogrify, then you can choose to only offer one constructor using the `SystemTypeTag` flavor, so that every subclass is required to declare its transmogrify-as type.  This would not be good advice by default, but might apply in certain cases.

> VectorSystem has both constructors but TimeVaryingAffineSystem has just the first one. AffineSystem has neither, suggesting it can't be extended(?). In that case LinearSystem (which derives from AffineSystem) can't be tranmogrified now. Is that right?

Well, the statement of the facts is right, but I'd summarize a different conclusion.  With this PR, LinearSystem can be transmogrified exactly as well as it could on master right now -- on master, AffineSystem's `DoTo..` methods are final, which means that LinearSystem does transmogrify, but ends up as, e.g., `AffineSystem<AutoDiffXd>`, with zero for the `y0` and `f0` terms (so is algebraically linear, but lots its C++ type that proved it so).  Since `LinearSystem`'s goal in life is just constructor sugar, this seems fine though a bit silly.  Certainly not a regression.

But, that the most derived type is lost during transmogrification _is_ weird, so perhaps I should take this opportunity to make LinearSystem transmogrify into a LinearSystem just so people don't get confused again.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-KqYw69IGpcYKNgOAZsu:b-ilrblm)*

**Comment by jwnimmer-tri** [comment]
Sigh, it goes even deeper.  `MatrixGain` is-a `LinearSystem` is-a `AffineSystem`, so I have to weave the transmogrification overloads all the way up the hierarchy.  Stay tuned for another push, a bit later on.

**Comment by jwnimmer-tri** [comment]
@sherm1 Back over to you, PTAL.  The question is if we like this design, where `SystemScalarConverter` is passed down from protected constructors, public constructors should not be used by subclasses, and terminal subclasses are final.  If so, I can work on updating the Doxygen to be clearer for everyday users, probably by including a separate tutorial section on why and how to support scalar conversion in Systems.

**Comment by sherm1** [comment]
This is a little awkward but I think good instructions could remedy that. I think those instructions should mandate the SystemTypeTag form of the constructor and we should stick with that in examples. I'm not convinced that we will need in practice the extra flexibility provided by creating your own SystemScalarConverter.

I'm not sure I understand why the terminal classes are final. They don't provide the infrastructure to allow further-derived classes to be transmogrified, but someone who didn't care about that should still be able to extend them I think. I suspect that out in the wild not everyone will be as enamored with transmogrifying as we are (not sure though). Can you say more about why you think these should be final?

A few comments below ...

---

Reviewed 14 of 14 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/simple_powertrain.h, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6nQ-AriOM3k_DNG4a:-Kr6nQ-AriOM3k_DNG4b:bjz315i) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/automotive/simple_powertrain.h#L41)):*
> ```objc
>       : systems::LinearSystem<T>(
>             systems::SystemScalarConverter{
>                 systems::SystemTypeTag<automotive::SimplePowertrain>{}},
> ```

BTW Clunky. Why not just 
`systems::LinearSystem<T>(systems::SystemTypeTag<SimplePowertrain>())`?

---

*[drake/systems/framework/test/vector_system_test.cc, line 478 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6p-QxTZkNvpxpFgCK:-Kr6p-QxTZkNvpxpFgCL:b-2txxb9) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/framework/test/vector_system_test.cc#L478)):*
> ```cpp
>     // This will fail at compile-time if T is symbolic::Expression.
>     bool condition = (T{0.0} == T{0.0});
>     (void)(condition);
> ```

BTW I don't understand this in several dimensions:
- why wouldn't two symbolic constant 0 expressions compare equal? If it fails now couldn't that be implemented at some point?
- why is this here at all?
- what's with the `(void)(condition)` line? Is that like `unused(condition)`?

Please clarify.

---

*[drake/systems/primitives/linear_system.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6ppCpo0sfk1BldDKe:-Kr6ppCpo0sfk1BldDKf:bphqyyr) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/primitives/linear_system.h#L88)):*
> ```objc
>                const Eigen::Ref<const Eigen::MatrixXd>& C,
>                const Eigen::Ref<const Eigen::MatrixXd>& D,
>                double time_period);
> ```

BTW I would like to see a `LinearSystem(SystemTypeTag,...)` protected constructor here so that derived types like SimplePowertrain don't have to use the clunky SystemScalarConverter boilerplate. We could also mandate that sugar in the instructions for writing a Drake-acceptable intermediate class.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Kr6vr22xTQ-ceYU-kKI:bndkfxi)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/primitives/linear_system.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6ppCpo0sfk1BldDKe:-Kr7gvwPH0ajPew6yyxh:b-atgcfp) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/primitives/linear_system.h#L88)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I would like to see a `LinearSystem(SystemTypeTag,...)` protected constructor here so that derived types like SimplePowertrain don't have to use the clunky SystemScalarConverter boilerplate. We could also mandate that sugar in the instructions for writing a Drake-acceptable intermediate class.
</blockquote></details>

This constructor is the only way to _disable_ transmogrification for subclasses that don't want it.  So I think the suggestion is to add _additional_ overload for a `SystemTypeTag`, so things like `SimplePowertrain` need less syntax.  Is that correct?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by jwnimmer-tri** [comment]
*[drake/systems/primitives/linear_system.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6ppCpo0sfk1BldDKe:-Kr7jj_K1Lx69heiBHbC:b-98sxh9) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/primitives/linear_system.h#L88)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

This constructor is the only way to _disable_ transmogrification for subclasses that don't want it.  So I think the suggestion is to add _additional_ overload for a `SystemTypeTag`, so things like `SimplePowertrain` need less syntax.  Is that correct?
</blockquote></details>

I guess the other option would be using something like `SystemTypeTag<void>{}` to disable transmogrification, instead of `SystemScalarConverter{}`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by sherm1** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/primitives/linear_system.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6ppCpo0sfk1BldDKe:-Kr7l3Ysq5W9g8wRsGSs:bw49xlw) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/primitives/linear_system.h#L88)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I guess the other option would be using something like `SystemTypeTag<void>{}` to disable transmogrification, instead of `SystemScalarConverter{}`.
</blockquote></details>

Yes, I am suggesting adding an additional overload if we *must* have the SystemScalarConverter() version. Great if we could get away with *only* the tag version -- `SystemTypeTag<void>` could be good, but I think just delegating to the public constructor that doesn't take a converter or tag would have the same effect?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/primitives/linear_system.h, line 88 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6ppCpo0sfk1BldDKe:-Kr7tmcJ0jNnW3I3btEk:b9vz2y7) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/primitives/linear_system.h#L88)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Yes, I am suggesting adding an additional overload if we *must* have the SystemScalarConverter() version. Great if we could get away with *only* the tag version -- `SystemTypeTag<void>` could be good, but I think just delegating to the public constructor that doesn't take a converter or tag would have the same effect?
</blockquote></details>

The public `LinearSystem(A, B, C, D, period)` constructor _does_ enable transmogrification (as a `LinearSystem`) -- its only allowed to be called by users, _not_ ever by derived classes, since it would transmogrify incorrectly.

I will explore adding `SystemTypeTag` support here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by jwnimmer-tri** [comment]
Review status: 13 of 17 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/test/vector_system_test.cc, line 478 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6p-QxTZkNvpxpFgCK:-Kr8355wQ-eIp0WPIIzq:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/systems/framework/test/vector_system_test.cc#L478)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW I don't understand this in several dimensions:
- why wouldn't two symbolic constant 0 expressions compare equal? If it fails now couldn't that be implemented at some point?
- why is this here at all?
- what's with the `(void)(condition)` line? Is that like `unused(condition)`?

Please clarify.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by sherm1** [comment]
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by jwnimmer-tri** [comment]
So it turns out that removing `explicit` from the `SystemScalarConverter` constructor (that takes a `SystemTypeTag`) makes so much syntax better.  I've pushed that here.

I still need to go back and rework a bunch of documentation, write a consolidated tutorial, and hunt down now-missing unit tests.

As always comments are welcome, but it's okay to wait until my next "PTAL" prompt to revisit the review, since this isn't fully cleaned-up yet.

---

Review status: 11 of 20 files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/simple_powertrain.h, line 41 at r2](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Kr6nQ-AriOM3k_DNG4a:-Kr8v-Ip-xVkis5cVXEW:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/729e639d1978095e642d655645ed1a94f312b283/drake/automotive/simple_powertrain.h#L41)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW Clunky. Why not just 
`systems::LinearSystem<T>(systems::SystemTypeTag<SimplePowertrain>())`?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Kr8uj2ZjKCXgZZgKFDU:b-8ovrb9)*

**Comment by sherm1** [comment]
Great idea -- looks much nicer!

---

Reviewed 9 of 9 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Kr91lrTTMzIQ8eSmwwF:b-pkv4os)*

**Comment by jwnimmer-tri** [comment]
FYI I've checkpointed the skeleton of new documentation, but still need to flesh out the actual Doxygen tutorial.  It's fine to delay review.

---

Review status: 13 of 31 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-KrInnLh84E4-P6t8nNe:b-5bv76i)*

**Comment by jwnimmer-tri** [comment]
This PR should finally be up-to-date again.  PTAL @sherm1 when you have time.

**Comment by sherm1** [comment]
:lgtm:

---

Reviewed 14 of 16 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5Bptk7uxQd0A6lq7p:bnfp4nl)*

**Comment by jwnimmer-tri** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5I8WjN4GKZrKbloo_:-Ku5I8WkePrsZBGqi-nu:b-wmu87j) (no related file):*
Needs platform review, I'll solicit help.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by sherm1** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5I8WjN4GKZrKbloo_:-Ku5KRwOSMaExwUL1BZJ:b-ws7yxo) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

Needs platform review, I'll solicit help.
</blockquote></details>

You can consider my lgtm for both if you want @jwnimmer-tri since this is mostly comments. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by ggould-tri** [comment]
+@ggould-tri 

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5PbM0xdUNhAuza0op:b-who1rq)*

**Comment by ggould-tri** [comment]
We seem to pass-by-value the `SystemScalarConverter` throughout, which is surprising in our pass-by-constref world.  

The only other nontrivial pass-by-value type I've seen in Drake, `drake::geometry::Identifier`, calls that out explicitly in its documentation: "It is the designed intent of this class, that ids derived from this class can be passed and returned by value. Passing ids by const reference should be considered a misuse."

If that's the intent of `SystemScalarConverter`then I propose that it get a similar clause for consistency and least-surprise.

Apart from that, and minor clarity issues below, :lgtm:

---

Reviewed 2 of 14 files at r2, 3 of 9 files at r4, 14 of 16 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[drake/systems/framework/leaf_system.h, line 250 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5QzgsSM5nJ0_GTRdh:-Ku5Qzgtl3LYivhGgTp0:b-12zmrw) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/leaf_system.h#L250)):*
> ```objc
> 
>   /// Constructor that declares no inputs, outputs, state, parameters, or
>   /// events, but allows subclasses declare scalar-type conversion support
> ```

FYI: Grammar.  "allows subclasses *to* declare"?

---

*[drake/systems/framework/leaf_system.h, line 254 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5Xd1kALa69fnmXcv8:-Ku5Xd1lXAdsguVD93x2:b-ne5z70) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/leaf_system.h#L254)):*
> ```objc
>   ///
>   /// The scalar-type conversion support will use @p converter.  To disable
>   /// scalar-type conversion support, pass a default-constructed @p converter.
> ```

FYI -- Way past rule-of-three on the previous two lines.  I'm not sure they are needed in any case:  This ctor is usable only by subclasses, and anyone writing a scalar-convertible subclass who hasn't ready `system_scalar_conversion` is doomed anyway.  So possibly the `See @ref` below is adequate for these sites.

---

*[drake/systems/framework/leaf_system.h, line 256 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5RiIMJep9yxvlP9Xq:-Ku5RiIMJep9yxvlP9Xr:b-nn4ijm) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/leaf_system.h#L256)):*
> ```objc
>   /// scalar-type conversion support, pass a default-constructed @p converter.
>   /// To enable scalar-type conversion support, pass a `SystemTypeTag<S>{}`
>   /// where `S` must be the most-derived concrete System subclass of `this`.
> ```

FYI -- This sentence is a maze of twisty little passages; it took me several attempts to figure it out.  Can it be rewritten more clearly?  Certainly the explanation in `system_scalar_conversion` is not so complicated, and just implies that `S` should be the subclass, full stop.

---

*[drake/systems/framework/vector_system.h, line 85 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5TeyXIfYUuklSyF8R:-Ku5TeyXIfYUuklSyF8S:ba1oo77) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/vector_system.h#L85)):*
> ```objc
>   ///
>   /// Does *not* declare scalar-type conversion support (AutoDiff, etc.).  To
>   /// enable AutoDiff support, use the SystemScalarConverter-based constructor.
> ```

BTW a one-line example here might save us some inquiries on slack #system_framework.

---

*[drake/systems/framework/test/vector_system_test.cc, line 476 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5UQoKe9GczDBQPlVe:-Ku5UQoKe9GczDBQPlVf:b-megkw7) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/test/vector_system_test.cc#L476)):*
> ```cpp
>   DirectScalarTypeConversionSystem()
>       : VectorSystem<T>(MakeConverter(), 0, 0) {
>     // This will fail at compile-time if T is ever symbolic::Expression.
> ```

Holy CRAP is this obscure.
Not defective, just baffling.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5YV78t_pSUUQJSt4f:bijij7z)*

**Comment by sherm1** [comment]
I think the cool C++ kids are switching to "prefer pass by value" from "prefer pass by const reference". Move semantics changes the balance here. See for example [this answer](https://stackoverflow.com/questions/2139224/how-to-pass-objects-to-functions-in-c/2139254#2139254) in stackoverflow.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5e6y1n_uKOmnZlDOi:bsg5594)*

**Comment by ggould-tri** [comment]
The cool kids are cool and all but perhaps we should update the style guide if we mean to be that cool.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5lVf_B3eL6Q_oGj_A:b-hp5g1x)*

**Comment by jwnimmer-tri** [comment]
For style guide changes, I can PR separately, but what would precisely be helpful?  I'm not sure that we should _require_ Drake code to use move semantics for taking ownership of moveable types.  Is the idea to find some way to make Drake developers more aware of this approach?  Or perhaps to bless that move-by-value an equally valid option as pass-by-const-ref?

---

Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/systems/framework/leaf_system.h, line 254 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5Xd1kALa69fnmXcv8:-Ku5uvPiVO0XMbpvryt0:b-b84vcl) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/leaf_system.h#L254)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- Way past rule-of-three on the previous two lines.  I'm not sure they are needed in any case:  This ctor is usable only by subclasses, and anyone writing a scalar-convertible subclass who hasn't ready `system_scalar_conversion` is doomed anyway.  So possibly the `See @ref` below is adequate for these sites.
</blockquote></details>

In other classes where this sentence is duplicated (the AffineSystem stuff), the public constructor says "subclasses must use the protected constructor", and the protected constructor contains this sentence as advice.  That's because the public constructor declares transmogrification -- declares that the most-derived subclass is itself.  The motivation for _those_ pastes of this comment is for the users who aren't doing transmogrification at all, so that they have an easy reference at hand for how to ignore the whole thing when they extend, e.g. LinearSystem.  For those cases, possibly I should write up an example of _ignoring_ the whole thing, and cite that example, instead of repeating this sentence?  What do you think?

For the paste here in LeafSystem (and in VectorSystem) -- for classes that themselves never declare intrinsic transmogrifcation support -- I can see how this sentence is not helpful, so I'll plan to remove it (or maybe reword it).

---

*[drake/systems/framework/vector_system.h, line 85 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5TeyXIfYUuklSyF8R:-Ku5mcZTHLwpd71AtjpL:b1n1yqe) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/vector_system.h#L85)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

BTW a one-line example here might save us some inquiries on slack #system_framework.
</blockquote></details>

I am not exactly sure what you mean.  Would a citation to the "Example using drake::systems::VectorSystem as the base class" in system_scalar_conversion_doxygen.h be the helpful thing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5u6GBdk01ylkYhirJ:b-wdlpf2)*

**Comment by ggould-tri** [comment]
pass-by-value is less google-disfavored than I recalled (possibly that's updated since I last read it?) so no defect.

---

Review status: 15 of 19 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/systems/framework/vector_system.h, line 85 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5TeyXIfYUuklSyF8R:-Ku5yOmAfx3IDuyAZsZd:b-s4z0aj) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/vector_system.h#L85)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary><blockquote>

I am not exactly sure what you mean.  Would a citation to the "Example using drake::systems::VectorSystem as the base class" in system_scalar_conversion_doxygen.h be the helpful thing?
</blockquote></details>

Yes, that would be fine.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5xvYrLtFqiL2cGQ_c:b-s8yb9f)*

**Comment by jwnimmer-tri** [comment]
*[drake/systems/framework/leaf_system.h, line 256 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5RiIMJep9yxvlP9Xq:-Ku5ywrz0bPEEjLSqyRr:b-m6p28e) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/leaf_system.h#L256)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

FYI -- This sentence is a maze of twisty little passages; it took me several attempts to figure it out.  Can it be rewritten more clearly?  Certainly the explanation in `system_scalar_conversion` is not so complicated, and just implies that `S` should be the subclass, full stop.
</blockquote></details>

Done, maybe?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by jwnimmer-tri** [comment]
I'm happy for no defect, but if there's anything we should still hint or clarify in our docs, I can take it on.

---

Reviewed 1 of 4 files at r6.
Review status: 15 of 19 files reviewed at latest revision, 1 unresolved discussion.

---

*[drake/systems/framework/vector_system.h, line 85 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-Ku5TeyXIfYUuklSyF8R:-Ku5zntUalY_Msxyw4H4:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5db141f75ed60868c4693bcfea5803008d543a72/drake/systems/framework/vector_system.h#L85)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary><blockquote>

Yes, that would be fine.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku5zpGYiYJy1W6w1Gla:bdfx46c)*

**Comment by ggould-tri** [comment]
Reviewed 1 of 4 files at r6.
Review status: 16 of 19 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743)*

**Comment by sherm1** [comment]
The basic idea of pass-by-value-then-move is more of an "effective use of C++11" strategy than a style issue. It's benefits are minor: possible copy avoidance, and an API with less ugly punctuation for parameters (`f(string)` beats `f(const string&)`) with downsides of trickier implementation and a possible note to callers promising not to make unnecessary copies, despite appearances.

A possible style guide entry might be prohibition, or at least discouragement, of taking large objects by value *unless* the code `std::move`s from them.

---

Review status: 16 of 19 files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/6743#-:-Ku63reKeb9ZWFBBQfRD:b-8y9wt3)*
